In [4]:
import pandas as pd


The objective of this analysis is to identify factors associated with loan defaults and provide recommendations that can help reduce credit risk

1. What is the overall loan default rate?
2. Does age affect the likelihood of default?
3. Does income affect the likelihood of default?
4. Which loan purposes have the highest default rates?
5. How does loan grade affect default risk?
6. Are larger loans associated with higher defaults?
7. Does previous default history affect future defaults?
8. Which borrower segments are highest risk?

In [6]:
df=pd.read_csv(r"C:\Users\Unnati\Downloads\archive (1)\credit_risk_dataset.csv")

In [8]:
df

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
0,22,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4
...,...,...,...,...,...,...,...,...,...,...,...,...
32576,57,53000,MORTGAGE,1.0,PERSONAL,C,5800,13.16,0,0.11,N,30
32577,54,120000,MORTGAGE,4.0,PERSONAL,A,17625,7.49,0,0.15,N,19
32578,65,76000,RENT,3.0,HOMEIMPROVEMENT,B,35000,10.99,1,0.46,N,28
32579,56,150000,MORTGAGE,5.0,PERSONAL,B,15000,11.48,0,0.10,N,26


In [16]:
df.shape
df.head()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   person_age                  32581 non-null  int64  
 1   person_income               32581 non-null  int64  
 2   person_home_ownership       32581 non-null  object 
 3   person_emp_length           31686 non-null  float64
 4   loan_intent                 32581 non-null  object 
 5   loan_grade                  32581 non-null  object 
 6   loan_amnt                   32581 non-null  int64  
 7   loan_int_rate               29465 non-null  float64
 8   loan_status                 32581 non-null  int64  
 9   loan_percent_income         32581 non-null  float64
 10  cb_person_default_on_file   32581 non-null  object 
 11  cb_person_cred_hist_length  32581 non-null  int64  
dtypes: float64(3), int64(5), object(4)
memory usage: 3.0+ MB


In [14]:
df.isnull().sum()
(df.isnull().sum() / len(df)) * 100

person_age                    0.000000
person_income                 0.000000
person_home_ownership         0.000000
person_emp_length             2.747000
loan_intent                   0.000000
loan_grade                    0.000000
loan_amnt                     0.000000
loan_int_rate                 9.563856
loan_status                   0.000000
loan_percent_income           0.000000
cb_person_default_on_file     0.000000
cb_person_cred_hist_length    0.000000
dtype: float64

In [34]:
clean_data=df.copy()
clean_data["person_emp_length"] = clean_data["person_emp_length"].fillna(df["person_emp_length"].median())
clean_data["loan_int_rate"]=clean_data["loan_int_rate"].fillna(df["loan_int_rate"].median())


In [38]:
clean_data.isnull().sum()   ##Handled the missing values.

person_age                    0
person_income                 0
person_home_ownership         0
person_emp_length             0
loan_intent                   0
loan_grade                    0
loan_amnt                     0
loan_int_rate                 0
loan_status                   0
loan_percent_income           0
cb_person_default_on_file     0
cb_person_cred_hist_length    0
dtype: int64

In [46]:
clean_data.duplicated().sum() ##Handling the duplicated values

165

In [48]:
clean_data.drop_duplicates(inplace=True)

In [75]:
clean_data.describe()

,person_age,person_income,person_emp_length,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_cred_hist_length
count,32411.000000,3.241100e+04,32411.000000,32411.000000,32411.000000,32411.000000,32411.000000,32411.000000
mean,27.730369,6.589796e+04,4.768720,9593.591682,11.014528,0.218722,0.170259,5.811052
std,6.210448,5.252093e+04,4.090478,6322.529723,3.083234,0.413386,0.106808,4.057816
min,20.000000,4.000000e+03,0.000000,500.000000,5.420000,0.000000,0.000000,2.000000
25%,23.000000,3.850000e+04,2.000000,5000.000000,8.490000,0.000000,0.090000,3.000000
50%,26.000000,5.500000e+04,4.000000,8000.000000,10.990000,0.000000,0.150000,4.000000
75%,30.000000,7.920000e+04,7.000000,12250.000000,13.110000,0.000000,0.230000,8.000000
max,94.000000,2.039784e+06,123.000000,35000.000000,23.220000,1.000000,0.830000,30.000000


In [71]:
clean_data[clean_data["person_age"] > 100]

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
81,144,250000,RENT,4.0,VENTURE,C,4800,13.57,0,0.02,N,3
183,144,200000,MORTGAGE,4.0,EDUCATION,B,6000,11.86,0,0.03,N,2
575,123,80004,RENT,2.0,EDUCATION,B,20400,10.25,0,0.25,N,3
747,123,78000,RENT,7.0,VENTURE,B,20000,10.99,0,0.26,N,4
32297,144,6000000,MORTGAGE,12.0,PERSONAL,C,5000,12.73,0,0.00,N,25


During data validation, a small number of records contained unrealistic ages (123 and 144 years). Since these values were likely data-entry errors and represented less than 0.1% of the dataset, the records were removed to improve data quality.

In [73]:
clean_data = clean_data[clean_data["person_age"] <= 100] ##we only keep person with age less than 100

In [77]:
clean_data[clean_data["person_emp_length"] > clean_data["person_age"]]

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
0,22,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3
210,21,192000,MORTGAGE,123.0,VENTURE,A,20000,6.54,0,0.10,N,4


In [79]:
clean_data=clean_data[clean_data["person_emp_length"] <= clean_data["person_age"]]

In [83]:
clean_data["person_emp_length"].max()

41.0

##Feature Engineering.
# Feature Engineering: Age Group
Created an `age_group` column by grouping borrowers into age ranges. This makes age-based default analysis easier and more meaningful than using individual ages.


In [89]:
clean_data["age_group"] = pd.cut(
    clean_data["person_age"],
    bins=[18, 25, 35, 45, 60, 100],
    labels=["18-25", "26-35", "36-45", "46-60", "60+"]
)

In [87]:
clean_data["age_group"].value_counts()

age_group
18-25    15243
26-35    13711
36-45     2809
46-60      581
60+         65
Name: count, dtype: int64

### Feature Engineering: Loan Size

Created a `loan_size` column by categorizing loan amounts into different size groups. This helps analyze whether larger loans are associated with higher default risk.

In [91]:
clean_data["loan_size"] = pd.cut(
    clean_data["loan_amnt"],
    bins=[0, 5000, 10000, 20000, 35000],
    labels=["Small", "Medium", "Large", "Very Large"]
)

In [93]:
clean_data["loan_amnt"].value_counts()

loan_amnt
10000    2649
5000     2032
12000    1795
6000     1794
15000    1496
         ... 
27300       1
17325       1
9625        1
4525        1
29850       1
Name: count, Length: 753, dtype: int64

In [95]:
clean_data["income_category"] = pd.qcut(
    clean_data["person_income"],
    q=4,
    labels=["Low", "Medium", "High", "Very High"]
)

In [106]:
clean_data["income_category"].value_counts()

income_category
Medium       8188
Low          8104
Very High    8099
High         8018
Name: count, dtype: int64

In [108]:
clean_data.to_csv("credit_risk_cleaned.csv", index=False)

In [112]:
#Establishing a connection between pyhton and database
import os
os.getcwd()

'C:\\Users\\Unnati\\Credit Risk Analysis Resume'

In [126]:
from sqlalchemy import create_engine

engine = create_engine(
    "mysql+pymysql://root:606%40Varun@localhost:3306/credit_risk_db"
)

In [131]:
clean_data.to_sql(
    name="credit_risk",
    con=engine,
    if_exists="replace",
    index=False
)

32409

In [134]:
from sqlalchemy import text
with engine.connect()as conn:
    result=conn.execute(text("SELECT 1"))
    print(result.scalar())

1
